Source/gendung.cpp

```diff
@@ -125,62 +110,43 @@ char dSpecial[MAXDUNX][MAXDUNY];
 int themeCount;
 THEME_LOC themeLoc[MAXTHEMES];
 
-void FillSolidBlockTbls()
+std::unique_ptr<byte[]> LoadLevelSOLData(size_t &tileCount)
 {
-	BYTE bv;
-	DWORD i, dwTiles;
-
-	memset(nBlockTable, 0, sizeof(nBlockTable));
-	memset(nSolidTable, 0, sizeof(nSolidTable));
-	memset(nTransTable, 0, sizeof(nTransTable));
-	memset(nMissileTable, 0, sizeof(nMissileTable));
-	memset(nTrapTable, 0, sizeof(nTrapTable));
-
-	std::unique_ptr<BYTE[]> pSBFile;
 	switch (leveltype) {
 	case DTYPE_TOWN:
 		if (gbIsHellfire)
-			pSBFile = LoadFileInMem("NLevels\\TownData\\Town.SOL", &dwTiles);
-		else
-			pSBFile = LoadFileInMem("Levels\\TownData\\Town.SOL", &dwTiles);
-		break;
+			return LoadFileInMem("NLevels\\TownData\\Town.SOL", &tileCount);
+		return LoadFileInMem("Levels\\TownData\\Town.SOL", &tileCount);
 	case DTYPE_CATHEDRAL:
 		if (currlevel < 17)
-			pSBFile = LoadFileInMem("Levels\\L1Data\\L1.SOL", &dwTiles);
-		else
-			pSBFile = LoadFileInMem("NLevels\\L5Data\\L5.SOL", &dwTiles);
-		break;
+			return LoadFileInMem("Levels\\L1Data\\L1.SOL", &tileCount);
+		return LoadFileInMem("NLevels\\L5Data\\L5.SOL", &tileCount);
 	case DTYPE_CATACOMBS:
-		pSBFile = LoadFileInMem("Levels\\L2Data\\L2.SOL", &dwTiles);
-		break;
+		return LoadFileInMem("Levels\\L2Data\\L2.SOL", &tileCount);
 	case DTYPE_CAVES:
 		if (currlevel < 17)
-			pSBFile = LoadFileInMem("Levels\\L3Data\\L3.SOL", &dwTiles);
-		else
-			pSBFile = LoadFileInMem("NLevels\\L6Data\\L6.SOL", &dwTiles);
-		break;
+			return LoadFileInMem("Levels\\L3Data\\L3.SOL", &tileCount);
+		return LoadFileInMem("NLevels\\L6Data\\L6.SOL", &tileCount);
 	case DTYPE_HELL:
-		pSBFile = LoadFileInMem("Levels\\L4Data\\L4.SOL", &dwTiles);
-		break;
+		return LoadFileInMem("Levels\\L4Data\\L4.SOL", &tileCount);
 	default:
 		app_fatal("FillSolidBlockTbls");
 	}
+}
 
-	const BYTE *pTmp = pSBFile.get();
-
-	for (i = 1; i <= dwTiles; i++) {
-		bv = *pTmp++;
-		if ((bv & 0x01) != 0)
-			nSolidTable[i] = true;
-		if ((bv & 0x02) != 0)
-			nBlockTable[i] = true;
-		if ((bv & 0x04) != 0)
-			nMissileTable[i] = true;
-		if ((bv & 0x08) != 0)
-			nTransTable[i] = true;
-		if ((bv & 0x80) != 0)
-			nTrapTable[i] = true;
-		block_lvid[i] = (bv & 0x70) >> 4; /* beta: (bv >> 4) & 7 */
+void FillSolidBlockTbls()
+{
+	size_t tileCount;
+	std::unique_ptr<byte[]> pSBFile = LoadLevelSOLData(tileCount);
+
+	for (unsigned i = 0; i < tileCount; i++) {
+		byte bv = pSBFile[i];
+		nSolidTable[i + 1] = (bv & byte { 0x01 }) != byte { 0 };
+		nBlockTable[i + 1] = (bv & byte { 0x02 }) != byte { 0 };
+		nMissileTable[i + 1] = (bv & byte { 0x04 }) != byte { 0 };
+		nTransTable[i + 1] = (bv & byte { 0x08 }) != byte { 0 };
+		nTrapTable[i + 1] = (bv & byte { 0x80 }) != byte { 0 };
+		block_lvid[i + 1] = static_cast<uint8_t>((bv & byte { 0x70 }) >> 4);

```

Comment:
Since you're casting at the end anyway:

```suggestion
		uint8_t bv = static_cast<uint8_t>(pSBFile[i]);
		nSolidTable[i + 1] = (bv & 0x01) != 0;
		nBlockTable[i + 1] = (bv & 0x02) != 0;
		nMissileTable[i + 1] = (bv & 0x04) != 0;
		nTransTable[i + 1] = (bv & 0x08) != 0;
		nTrapTable[i + 1] = (bv & 0x80) != 0;
		block_lvid[i + 1] = (bv & 0x70) >> 4;
```

Reply:
in that case it would be better to just load the file as uint8_t and avoid any casting.

Reply:
really wish bit-wise operations could be done against hex values. my main argument for doing it way i did is that we do not have a numeric value before doing the bit shift

Reply:
You can do bitwise operations with hex values no problem - std::byte is the weirdo here.

It's implemented as enum class explicitly to disallow any usual integer propagation, and it has only certain operators implemented (the bitwise ones, and only for operating with other byte values).

The hex values you typed in are integer literals, so to do the bitwise operation, the std::byte would have to be promoted to integer first - which is prohibited.

Reply:
Oh - adding operators that allow bitwise operations (or really pretty much any operation) with other data types on std::byte isn't miracle work, but it softens up the strict typing even more. At that point, why not just use BYTE/unsigned char/uint8_t.